test_utils/src/cluster.rs

```diff
@@ -0,0 +1,116 @@
+//! This module provides test functionality to build cluster objects from (YAML) files and matches
+//! contained data definitions for testing.
+//! Each cluster objects consists of a Load trait and a Data trait.
+//! The goal is to provide interchangeable cluster objects for testing.
+use serde::de;
+
+/// Loads a cluster object and parses the content to T
+pub trait Load<T>
+where
+    T: de::DeserializeOwned,
+{
+    type Cluster;
+
+    fn load() -> T;
+}
+
+/// Contains object definitions that should match the provided file data
+pub trait Data {
+    const MASTER_SELECTOR_1_NODE_NAME: &'static str;
+    const MASTER_SELECTOR_1_INSTANCES: usize;
+    const MASTER_SELECTOR_1_PORT: usize;
+    const MASTER_SELECTOR_1_WEB_UI_PORT: usize;
+    const MASTER_SELECTOR_1_CONFIG_PORT: usize;
+    const MASTER_SELECTOR_1_ENV_PORT: usize;
+
+    const MASTER_SELECTOR_2_NODE_NAME: &'static str;
+    const MASTER_SELECTOR_2_INSTANCES: usize;
+    const MASTER_SELECTOR_2_PORT: usize;
+    const MASTER_SELECTOR_2_WEB_UI_PORT: usize;
+    const MASTER_SELECTOR_2_ENV_PORT: usize;
+
+    const MASTER_SELECTOR_3_NODE_NAME: &'static str;
+    const MASTER_SELECTOR_3_INSTANCES: usize;
+    const MASTER_SELECTOR_3_PORT: usize;
+
+    const MASTER_SELECTOR_4_NODE_NAME: &'static str;
+    const MASTER_SELECTOR_4_INSTANCES: usize;
+
+    const WORKER_SELECTOR_1_NODE_NAME: &'static str;
+    const WORKER_SELECTOR_1_INSTANCES: usize;
+    const WORKER_SELECTOR_1_CORES: usize;
+    const WORKER_SELECTOR_1_MEMORY: &'static str;
+    const WORKER_SELECTOR_1_ENV_MEMORY: &'static str;
+
+    const WORKER_SELECTOR_2_NODE_NAME: &'static str;
+    const WORKER_SELECTOR_2_INSTANCES: usize;
+    const WORKER_SELECTOR_2_CORES: usize;
+    const WORKER_SELECTOR_2_MEMORY: &'static str;
+
+    const HISTORY_SERVER_SELECTOR_2_NODE_NAME: &'static str;
+    const HISTORY_SERVER_SELECTOR_2_INSTANCES: usize;
+    const HISTORY_SERVER_SELECTOR_2_CORES: usize;
+    const HISTORY_SERVER_SELECTOR_2_MEMORY: &'static str;
+
+    const CLUSTER_VERSION: &'static str;
+    const CLUSTER_SECRET: &'static str;
+    const CLUSTER_LOG_DIR: &'static str;
+    const CLUSTER_MAX_PORT_RETRIES: usize;
+}
+
+pub struct TestSparkCluster;
+
+impl<T> Load<T> for TestSparkCluster
+where
+    T: de::DeserializeOwned,
+{
+    type Cluster = T;
+
+    fn load() -> T {
+        let file = include_str!("../data/test_spark_cluster.yaml");
+        serde_yaml::from_str(file).unwrap()
+    }
+}
+
+impl Data for TestSparkCluster {
```

Comment:
Isn't this what you wanted to avoid? Now you specify things in two different locations.
You don't have to change it but I'm not a big fan.

Reply:
Yes you still need to adapt in 2 places if you change anything. I would prefer to just build up the test cluster here and not uses files at all. That would result in only one place to change/adapt. However i tried that and as far as i understood you cannot have circle dependencies. E.g. test_utils depending on crd to get the cluster definition and then crd depending on test_utils to test the cluster...I found some workarounds but i did not like any of them so i picked that apporach for now.